src/cppext.cpp

```diff
@@ -313,6 +313,207 @@ cdd cdd_past(const cdd& state)
+     * of the current row are copied to the next row up to and including the provided column.
+     * @param copy_to_column copy the previous values up to this column index (including this column).
+     */
+    void next_row(int copy_to_column)
+    {
+        assert(copy_to_column < num_cols);
+        assert(!matrix.empty());
+
+        std::vector<int32_t> new_row;
+
+        // Copy values from the previous row.
+        for (int j = 0; j <= copy_to_column; j++) {
+            new_row.push_back(matrix.at(matrix.size() - 1).at(j));
+        }
+        matrix.push_back(new_row);
+        rows_to_be_ignored.push_back(false);
+    }
+
+    /**
+     * Indicate that the current row needs to be ignored.
+     */
+    void ignore_current_row() { rows_to_be_ignored.back() = true; }
+
+    /**
+     * Delete the ignored rows from the matrix and shift the remaining rows upwards, i.e.,
+     * if row \a is to be ignored, the new row \i of the matrix is row \i + 1;
+     */
+    void delete_ignored_rows()
+    {
+        int i = 0;
+        for (auto it = matrix.begin(); it != matrix.end();) {
+            if (rows_to_be_ignored.at(i)) {
+                matrix.erase(it);
+            } else {
+                ++it;
+            }
+            ++i;
+        }
+    }
+
+    /**
+     * Get the current matrix as a single dimensional array of size \a num_cols * (\a current_row + 1).
+     * Entries that were not added will get the default value.
+     * @return the matrix as a single row array.
+     */
+    int32_t* get_array(int32_t default_value)
+    {
+        auto* array = new int32_t[num_cols * (matrix.size())];
+
+        for (int i = 0; i < matrix.size(); i++) {
+            int j = 0;
+            for (; j < matrix.at(i).size(); j++) {
+                array[i * num_cols + j] = matrix.at(i).at(j);
+            }
+            for (; j < num_cols; j++) {
+                array[i * num_cols + j] = default_value;
+            }
+        }
+
+        return array;
+    }
+
+    /**
+     * Get the current row size of the matrix.
+     * @return the current row size.
+     */
+    uint32_t get_current_row_size() { return matrix.size(); }
+
+private:
+    std::vector<std::vector<int32_t>> matrix;
+    std::vector<bool> rows_to_be_ignored;
+    int num_cols;
+};
+
+/**
+ * Transform a BDD recursively into an matrix representation.
+ *
+ * @param r a cdd node from the bdd.
+ * @param varsMatrix the matrix containing the traces with the boolean variable id numbers.
+ * @param valuesMatrix the matrix containing the traces with the boolean variable values.
+ * @param current_step the current step number (= depth of the trace from the top BDD node).
+ * @param negated Whether this node is reached by (an odd number of) negated node(s).
+ * @param num_bools the number of boolean variables.
+ */
+void cdd_bdd_to_matrix_rec(ddNode* r, dynamic_two_dim_matrix* varsMatrix, dynamic_two_dim_matrix* valuesMatrix,
```

Comment:
In C++ context it would be better to pass-by-reference instead of pass-by-pointer.